mixer/adapter/signalfx/registry.go

```diff
@@ -0,0 +1,191 @@
+// limitations under the License.
+
+package signalfx
+
+import (
+	"container/list"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/signalfx/golib/datapoint"
+	"github.com/signalfx/golib/sfxclient"
+)
+
+type metricID string
+
+var currentTime = time.Now
+
+// registry holds a set of collectors that produce datapoints on demand.  This
+// allows us to smooth out the metrics we get to make them regular timeseries.
+// We also need to expire collectors that haven't been updated after a certain
+// amount of time to prevent excessive memory usage.
+type registry struct {
+	cumulatives    map[metricID]*cumulativeCollector
+	rollingBuckets map[metricID]*sfxclient.RollingBucket
+
+	// A linked list that we keep sorted by access time so that we can very
+	// quickly tell which collectors are expired and should be deleted.
+	lastAccessList list.List
+	// A map optimizing lookup of the access time elements that are used in the
+	// above linked list.
+	lastAccesses map[metricID]*list.Element
+
+	expiryTimeout time.Duration
+	lock          sync.RWMutex
+}
+
+var _ sfxclient.Collector = &registry{}
+
+func newRegistry(expiryTimeout time.Duration) *registry {
+	return &registry{
+		cumulatives:    make(map[metricID]*cumulativeCollector),
+		rollingBuckets: make(map[metricID]*sfxclient.RollingBucket),
+		lastAccesses:   make(map[metricID]*list.Element),
+		expiryTimeout:  expiryTimeout,
+	}
+}
+
+func (r *registry) RegisterOrGetCumulative(name string, dims map[string]string) *cumulativeCollector {
+	id := idForMetric(name, dims)
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if c := r.cumulatives[id]; c == nil {
+		r.cumulatives[id] = &cumulativeCollector{
+			MetricName: name,
+			Dimensions: dims,
+		}
+	}
+
+	r.markUsed(id)
+	return r.cumulatives[id]
+}
+
+func (r *registry) RegisterOrGetRollingBucket(name string, dims map[string]string, intervalSeconds uint32) *sfxclient.RollingBucket {
+	id := idForMetric(name, dims)
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if c := r.rollingBuckets[id]; c == nil {
+		r.rollingBuckets[id] = sfxclient.NewRollingBucket(name, dims)
+		r.rollingBuckets[id].BucketWidth = time.Duration(intervalSeconds) * time.Second
+	}
+
+	r.markUsed(id)
+	return r.rollingBuckets[id]
+}
+
+func (r *registry) Datapoints() []*datapoint.Datapoint {
+	r.lock.RLock()
+
+	var out []*datapoint.Datapoint
+
+	for id := range r.cumulatives {
+		out = append(out, r.cumulatives[id].Datapoints()...)
+	}
+
+	for id := range r.rollingBuckets {
+		out = append(out, r.rollingBuckets[id].Datapoints()...)
+	}
+
+	r.lock.RUnlock()
+
+	r.purgeOldCollectors()
+
+	return out
+}
+
+type access struct {
+	ts time.Time
+	id metricID
+}
+
+// makeUsed should be called to indicate that a metricID has been accessed and
+// is still in use.  This causes it to move to the front of the lastAccessList
+// list with an updated access timestamp.
+// The registry lock should be held when calling this method.
```

Comment:
should we just have this be explicit in the method itself (like `purgeOldCollectors()` below)?

Reply:
This method is part of a larger critical section formed by the `RegisterOrGet*` methods and should not reacquire the lock separately or else there will be a race condition between purging collectors and marking them used.